File: beam-series-python/model_guide/ptransform_pardo.py
import logging
import apache_beam as beam
from apache_beam.options.pipeline_options import PipelineOptions

accounts = [
        {
            "name": "Lucas",
            "age": 32,
            "gender": "M",
            "email": "lucas@foo.com",
            "is_active": True
            },
        {
            "name": "Fernando",
            "age": 38,
            "gender": "M",
            "email": "fernando@foo.com",
            "is_active": True
            },
        {
            "name": "Maria",
            "age": 14,
            "gender": "F",
            "email": "maria@foo.com",
            "is_active": False
            },
        ]

def age_to_str(account: dict):
    account["age"] = str(account["age"])
    return account

class MyDoFn(beam.DoFn):

    def start_bundle(self):
        print("iniciando bundle")
    def process(self,element):
        if element["is_active"]:
            element["status"] = "approved"
        else:
            element["status"] = "rejected"
        yield element
    def finish_bundle(self):
        print("finalizando bundle")
def run():
    pipeline_options = PipelineOptions()
    with beam.Pipeline(options=pipeline_options) as p:

        pcoll = p | 'Create pcoll' >> beam.Create(accounts)

        beam_filter = pcoll | "Filter aged" >> beam.Filter(lambda account: account["age"] >= 18)
        beam_format = pcoll | "Format age" >> beam.Map(lambda account: age_to_str(account))
        beam_extract = pcoll | "extract email" >> beam.Map(lambda account: {"email": account["email"]})
        beam_custom = pcoll | "pardo" >> beam.ParDo(MyDoFn())
        #beam_filter | "Print filter" >> beam.Map(print)
        #beam_format | "Print format" >> beam.Map(print)
        #beam_extract | "Print email" >> beam.Map(print)
        beam_custom | "Print pardo" >> beam.Map(print)
        p.run()

if __name__ == '__main__':
    logging.getLogger().setLevel(logging.INFO)
    run()

File: beam-series-scala/src/main/scala/com/beam_series/beam_series_scala/model_guide/ptransformParDo.scala

package com.beam_series.beam_series_scala.model_guide
import com.spotify.scio._
import com.spotify.scio.values.SCollection
import org.apache.beam.sdk.transforms.DoFn._
import org.apache.beam.sdk.transforms.{DoFn, ParDo}
object ptransformParDo {

  class MyDoFn extends DoFn[Map[String,Any],Map[String,Any]] {

    @StartBundle
    def startBundle(c: DoFn[Map[String,Any],Map[String,Any]]#StartBundleContext): Unit = {
      println("Iniciando Bundle....")
    }
    @ProcessElement
    def processElement(@Element element: Map[String,Any], out: OutputReceiver[Map[String,Any]]): Unit = {
      if (element("is_active").asInstanceOf[Boolean] == true) {
        out.output(Map("account" -> element,"status" -> "approved"))
      } else {
        out.output(Map("account" -> element,"status" -> "rejected"))
      }
    }
    
    @FinishBundle
    def finishBundle(c: DoFn[Map[String,Any],Map[String,Any]]#FinishBundleContext): Unit = {
      println("Finalizando Bundle....")
    }
    
  }
  def main(cmdlineArgs: Array[String]): Unit = {

    val accounts: Seq[Map[String,Any]] = Seq(
      Map( 
        "name" -> "Lucas",
        "age" -> 32,
        "gender" -> "M",
        "email" -> "lucas@foo.com",
        "is_active" -> true
        ),
      Map( 
        "name" -> "Fernando",
        "age" -> 38,
        "gender" -> "M",
        "email" -> "fernando@foo.com",
        "is_active" -> true
        ),
      Map( 
        "name" -> "Maria",
        "age" -> 14,
        "gender" -> "F",
        "email" -> "maria@foo.com",
        "is_active" -> false
        )
      )
    val (sc, args) = ContextAndArgs(cmdlineArgs)


    val scoll = sc.parallelize(accounts)

    scoll
      //.filter(x => x("age").asInstanceOf[Int] >= 18)
      //.map(x=> x("email").asInstanceOf[String]) 
      .applyTransform(ParDo.of(new MyDoFn))
      .map(x => println(x))
    
    sc.run()

  }
}
